mlflow/pyfunc/__init__.py

```diff
@@ -828,13 +828,82 @@ def _create_model_downloading_tmp_dir(should_use_nfs):
 _MLFLOW_SERVER_OUTPUT_TAIL_LINES_TO_KEEP = 200
 
 
+def _convert_model_output_spec_to_spark_type(spec):
+    from mlflow.types.schema import ColSpec, TensorSpec, DataType
+    from pyspark.sql.types import ArrayType
+
+    # TODO: handle optional output columns.
+    if isinstance(spec, ColSpec):
+        return spec.type.to_spark()
+    elif isinstance(spec, TensorSpec):
+        data_type = DataType.from_numpy_type(spec.type)
+        if data_type is None:
+            raise MlflowException(
+                f"Model output tensor spec type {spec.type} is not supported in spark_udf."
+            )
+
+        if len(spec.shape) == 1:
+            return ArrayType(data_type.to_spark())
+        elif len(spec.shape) == 2:
+            return ArrayType(ArrayType(data_type.to_spark()))
+        else:
+            raise MlflowException(
+                "Only one dimension or 2 dimensions tensors are supported as spark_udf "
+                f"return value, but model output '{spec.name}' has shape {spec.shape}."
+            )
+    else:
+        raise ValueError(f"Unknown schema output spec {spec}.")
+
+
+def _infer_spark_udf_return_type(model_output_schema):
+    from pyspark.sql.types import StructType, StructField, ArrayType
+
+    if len(model_output_schema.inputs) == 1:
+        return _convert_model_output_spec_to_spark_type(model_output_schema.inputs[0])
+
+    return StructType(
+        [
+            StructField(
+                name=spec.name or str(i), dataType=_convert_model_output_spec_to_spark_type(spec)
+            )
+            for i, spec in enumerate(model_output_schema.inputs)
+        ]
+    )
+
+
 def _parse_spark_datatype(datatype: str):
     from pyspark.sql.functions import udf
 
     return udf(lambda x: x, returnType=datatype).returnType
 
 
-def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LOCAL):
+def _convert_array_values(values, spark_type, spark_primitive_type_to_np_type):
```

Comment:
Can we add type hints here for readability?

Reply:
I will add docstring. But is type hint required ? It will cause pyspark.sql package imported when import this module, but we should avoid it.

Reply:
No need